gaussian smooth: expose spatial smoothing in API

kecnry opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This pull request exposes spatial (in addition to spectral) smoothing in the user API for the gaussian smooth plugin, introduced in #1401, for cubeviz.

This PR *renames* the `apply_spectral_smooth` to `spectral_smooth`, and exposes two new methods:
* `apply_smooth(add_data=True)` (which listens to the `mode` and the results label component)
* `spatial_smooth()` (only exposed for cubeviz, just returns a spectrum1D object)
* `spectral_smooth()` (just returns a spectrum1D object)

Since #1401 has not been included in a release, this does not need to be considered breaking changes (as long as we get this in before the next release).

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->


### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

camipacifici commented:
If I understand correctly 
```
smoothing = viz.plugins['Gaussian Smooth']
smoothing.dataset = "whatever"
smoothing.mode = 'Spectral'
smoothing.stddev = 5
smoothing.apply_smooth(add_data=True)
```
and
```
smoothing = viz.plugins['Gaussian Smooth']
smoothing.dataset = "whatever"
smoothing.stddev = 5
smoothing.spectral_smooth()
```
do the same thing, but the first updates the entries in the GUI and shows the result, while the second is more compact but does not update the GUI or shows the result there.
My personal vote is to NOT expose `smoothing.spectral_smooth` and `smoothing.spatial_smooth` and give only the first option to the user. Happy to be convinced otherwise if people can come up with a good case.

kecnry commented:
Ok, it is easier to re-introduce later than to hide later, so I removed the `spectral_smooth` and `spatial_smooth` methods from the user API (while keeping the internal code refactor).  I've also renamed `apply_smooth` to `smooth` to be more consistent with the collapse plugin's `collapse` method.

I'm also not sure I love the name of the `mode` option, but couldn't think of anything clearer that is still concise.  The collapse plugin will probably also eventually have a spatial vs spectral switch, and we would probably want to use the same pattern there.  The collapse plugin currently uses `method` for mean/max/sum/etc, so we shouldn't use that here for spatial vs spectral (although maybe we should consider renaming `collapse.method` to `collapse.function` to be more consistent with the `plot_options.collapse_function`).

So..... how would everyone feel about the following additional changes:
* `gaussian_smooth.mode` -> `gaussian_smooth.along_axis` (and plan to do the same for `collapse` in the future)
* `collapse.method` -> `collapse.function` or `collapse.collapse_function`

PatrickOgle commented:
Can we instead make this implicit in the API:
smoothing.spectral.stddev = 5  invokes a spectral smoothing of 5
smoothing.spatial = 5 invokes a spatial smoothing of 5

PatrickOgle commented:
I am not sure collapse.function is any less ambiguous than collapse.method 
Furthermore, both are python terms that could be confusing.  I can't think of a good alternative at the moment, though...

pllim commented:
Code wise, LGTM, but looks like POs have concerns about the API, so I'll withhold technical review until that is settled.

kecnry commented:
@pllim - I think the API discussion is settled, at least for now (please anyone correct me if I'm wrong, but I think the general suggestion was to stick with `mode` here and to change `method` to `function` in #1702).  We may have to iterate further in the future, but this is at least a step in the right direction that I think we should try to get in before the next release to avoid absolutely needing to deprecate the API currently in main.

camipacifici reviewed (approved):
Looks good from PO perspective.
